### DISCUSSION TITLE
ComputeQpResidual in porous flow module

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#26371] [Q&A Modules: Porous Flow]
Dear all,

When I looked at your code in porousflow I found that there are two ways to write the ComputeQpResidual in Moose. 

1. the general one is based on [_qp] which can be easily understood that we integrate this term in the quadrature points.
 ```
------------------------------------------------------------------------------------------------------------------
ComputeQpResidual
return (*_relative_permeability)[_qp][_ph] * (*_fluid_density_node)[_qp][_ph] /(*_fluid_viscosity)[_qp][_ph] 
------------------------------------------------------------------------------------------------------------------
```
3. based on _i which is stated in your tutorial: index for test function. I understand it in the following as node number. Are the index of test function and node number same? So in this residual, we integrate the term only in the node points? then why we have Qp in the name of ComputeQpResidual?
```
------------------------------------------------------------------------------------------------------------------
ComputeQpResidual
return (*_relative_permeability)[_i][_ph] * (*_fluid_density_node)[_i][_ph] /(*_fluid_viscosity)[_i][_ph]
------------------------------------------------------------------------------------------------------------------
```
Thanks a lot.
Qia

**Comment by GiudGiud** [comment]
@cpgr 
I m not aware of this `_i` indexing in kernel-loops

**Reply by lindsayad** [reply]
For a Lagrange finite element family with order equal to the geometric element order, then `_i` will correspond to the local node numbers

**Reply by Oops-Qiao** [reply]
Thanks a lot.
My main question is 
if we integrate the node values in the residual, will it give the same or similar physical results as the integration of quadrature point values? How does it work in the ComputeQpResidual? I guess there should be some qp in the original code to define the ComputeQpResidual.

**Reply by lindsayad** [reply]
This could be related to [lumping](https://mooseframework.inl.gov/modules/porous_flow/mass_lumping.html) but we'll let @cpgr give the authoritative answer
